content/post/securing-stateless-apis.md

```diff
@@ -0,0 +1,114 @@
+malicious piece of code that can be executed on the server. Thus, data integrity is what
+needs to be maintained. In our case to check for the payload integrity we calculate the
+hash of the entire request payload by some secret which is a combination of a secret key
+and current timestamp.The key and the hash function is pre decided and known to both the
+client and the server. The hash value and the timestamp are sent over as request headers
+of the REST API. On receiving the request on the server, the server calculates the hash of
+the received request payload and matches it with the hash received in the request header.
+If the hash values do not match the request should be dropped without processing further.
+
+_Below are the steps to generate hash_:
+- Since we have the request payload in JSON format we convert it to a string.
+- Generate a hash value of the request payload string say `hash(client)` with the help of a
+  secret and a pre-decided hash function (HMAC/ECDSA), known to both server and client.
+- Add this hash value as one of your request headers. On receiving the request at the
+  server you can follow the similar steps as were followed at the client to create a hash
+  value say `hash(server)`.
+- Compare `hash(client)` and `hash(server)`. If both the hash values match let the server proceed
+  with the request else drop the request.
+
+## Broken Authentication and Authorization
+
+_Can you imagine an authentication/authorization mechanism without a database or an user
+model?_
+
+The answer is yes, but as we said we do not have any user model in place for our system.
+But in our case there is another server say Serveruser which is hosting all the user data.
+This Serveruser creates unique tokens for every user with some expiry time associated with
+it. So our server is accepting this token with each request from client and asks
+Serveruser to validate it. If Serveruer says the user is authenticated and authorized for
+the operation we proceed with the request. Also if token is expired our client will prompt
+user and ask for valid credentials to generate a new token. This new token will be passed
+with subsequent requests.
+
+## Man In The Middle (MITM) attacks
+
+Here data at rest was not our concern since we dont store any users data. But to secure
+data in motion we ensured data tansfer between our client and server along with to and
+from our server to ABC server if through secure network. This is achieved by using HTTPS
+channel secured by TLS protocol instead of simple HTTP. needs rephrasing
+
+## Replay attacks
+
+[Replay attacks](https://en.wikipedia.org/wiki/Replay_attack) are a broad term, but here
+it basically means to re-send any API request and get back the same request. To avoid such
+attacks we keep a request acceptance timeout  at our server, say 300 seconds. Steps to
+follow are: Now we will add client utc time in request header say timeclient. Server will
```

Comment:
Steps to follow are can be converted to _Steps to follow are_: and then further points we can convert to bullets